### PR TITLE
Learning curve charts in notebook live updated during training

### PR DESCRIPTION
**Pull request by leopd**
New model training callbacks PandasLogger and LiveLearningCurve.
PandasLogger builds a Dataframe with all training metrics.
LiveLearningCurve renders a chart in a notebook that updates live during training.
A little refactoring of callbacks from model.fit() to clean up and make this possible.

### CODE REVIEW

**Comment by piiswrong** [comment]
Cool!
Could you add this to mx.module?
Also an example showing usage would be helpful

**Comment by piiswrong** [comment]
Also would it better to make LiveBokehChart a internal class of PandasLogger?
Having something that's not a callback in mx.callback can be confusion

**Comment by leopd** [comment]
I've got some examples I can put in.  Adding it to module should be pretty straightforward.

I think PandasLogger & LiveBokehChart belong separate for a couple reasons:
1) I think it'll be useful to attach multiple live charts onto the same training run -- e.g. one showing speed, another showing loss, another showing validation metrics.  Or in some cases we track lots of different validation metrics.  This is easy to do through composition the way the classes are structured -- you instantiate multiple BokehChart objects and connect them to the same PandasLogger.  I imagine there should be other LiveBokehChart sub-classes that make these patterns easy.
2) The PandasLogger is likely to be useful by itself without any charts.  You can run this without a notebook, and it's trivial (one line of code) to save the dataframe as a .csv for offline analysis of the run.

That said, I think we can simplify the common case by having the BokehChart instantiate its own PandasLogger if one isn't passed in during construction.

**Comment by piiswrong** [comment]
Ok. Sounds good

File: python/mxnet/model.py [hunk 1]
```diff
@@ -112,13 +112,40 @@ def _update_params(param_arrays, grad_arrays, updater, num_device,
...
+
+
+class LogValidationMetricsCallback(object):
+    """Just logs the eval metrics at the end of an epoch.
+    """
+
+    def __call__(self,param):
+        if not param.eval_metric:
+            return
+        name_value = eval_metric.get_name_value()
```
Comment (mli) [inline python/mxnet/model.py]: should be `name_value = param.eval_metric.get_name_value()`

File: python/mxnet/model.py [hunk 2]
```diff
@@ -112,13 +112,40 @@ def _update_params(param_arrays, grad_arrays, updater, num_device,
...
+        for name, value in name_value:
+            logger.info('Epoch[%d] Validation-%s=%f', epoch, name, value)
```
Comment (mli) [inline python/mxnet/model.py]: logger is not available. we also need to change epoch to `param.epoch`

**Comment by mli** [comment]
the live board works nice. here is my codes
```python
pandas_logger = mx.callback.PandasLogger(batch_size, 5)
live = mx.callback.LiveLearningCurve(pandas_logger, 'accuracy', 5)
```
and add the following into `model.fit`
```
    **mx.callback.args_wrapper(pandas_logger, live)
```

but i feel it is too complicate, for example, i need to specify 5 twice, and even for `eval_metric='acc'`, i need to specify `'accuracy'` when creating  `LiveLearningCurve` 

i also changed the figure style to make it slightly better, namely

```python
        self.fig.circle(self.x1,self.y1, size=1.5, line_alpha=.2, legend="train")
        self.x2 = []
        self.y2 = []
        self.fig.line(self.x2,self.y2, line_color='green', line_width=2, legend="validation")
        self.fig.legend.location = "bottom_right"
```

here is screenshot
screen shot 2016-10-23 at 4 29 18 pm

**Comment by leopd** [comment]
Thanks for catching my bugs and the visual improvements, Mu.  I've got some more examples and a Module patch I'm about to merge in -- I'll add these.

**Comment by piiswrong** [comment]
We probably want a checkbox to toggle between wall clock and number of epochs because
1. people have different hardware setup with different speed, which makes results non comparable.
2. systems sometimes slowdown randomly and could be confusing if reflected on training curve

**Comment by leopd** [comment]
@piiswrong a constructor switch between epoch and wall-clock is easy to do.  I'll add that.  I like wall-clock for comparing different hyper-parameter settings though because what you really care about is how long it actually takes to get an answer.  It would be awesome to have a GUI checkbox that toggles, but that's actually fairly complicated due to a limitation in bokeh that doesn't let you dynamically swap axis ranges at runtime.

@mli Yes the constructor is too complicated right now.  I'm going to make the LiveLearningCurve instantiate a PandasLogger by default if you don't specify one.  This makes the common case simpler.  If you want multiple live charts, you'll be able to do that more efficiently by sharing a pandaslogger.  I'll push the new proposed interface soon and would love feedback.

**Comment by leopd** [comment]
I'm passing this on to one of my colleagues to clean up and finish.  He'll submit a new PR soon.

**Comment by leopd** [comment]
The new PR for this is https://github.com/dmlc/mxnet/pull/3800
